subsys/bluetooth/services/mds.c

```diff
@@ -0,0 +1,544 @@
+{
+	static const size_t att_header_length = 0x03;
+
+	size_t length;
+
+	if (!conn) {
+		return 0;
+	}
+
+	length = bt_gatt_get_mtu(conn);
+	if (length < (att_header_length + sizeof(struct mds_data_export_nfy))) {
+		LOG_ERR("MTU value too low: %d or link is disconnected", length);
+		return 0;
+	}
+
+	/* According to BLE Core v5.3 Vol 3, Part F 3.4.7.1 maximum supported length of the
+	 * notification is (ATT_MTU - 3).
+	 */
+	length -= att_header_length;
+	length -= sizeof(struct mds_data_export_nfy);
+
+	return length;
+}
+
+static int chunk_send(struct bt_conn *conn, struct net_buf_simple *buf)
+{
+	struct bt_gatt_notify_params params = {0};
+	static struct bt_gatt_attr *attr;
+
+	__ASSERT(conn, "Invalid parameters");
+	__ASSERT(buf, "Invalid parameters");
+
+	if (!attr) {
+		attr = bt_gatt_find_by_uuid(mds_svc.attrs, mds_svc.attr_count,
+					    BT_UUID_MDS_DATA_EXPORT);
+	}
+
+	__ASSERT_NO_MSG(attr);
+
+	params.attr = attr;
+	params.data = buf->data;
+	params.len = buf->len;
+	params.func = mds_sent_cb;
+
+	return bt_gatt_notify_cb(conn, &params);
+}
+
+static uint8_t chunk_number_update(uint8_t chunk_number)
+{
+	return ++chunk_number & MDS_CHUNK_NUMBER_MAX_VALUE;
+}
+
+static int mds_data_send(struct bt_conn *conn)
+{
+	int err;
+	bool data_available;
+	struct mds_data_export_nfy *data_export_nfy;
+	size_t chunk_max_size = chunk_data_length_get(mds_instance.conn);
+	size_t chunk_size = chunk_max_size;
+
+	NET_BUF_SIMPLE_DEFINE(buf, (sizeof(struct mds_data_export_nfy) + chunk_max_size));
+
+	data_export_nfy = net_buf_simple_add(&buf,
+					sizeof(struct mds_data_export_nfy) + chunk_max_size);
+	if (!data_export_nfy) {
+		LOG_ERR("Cannot allocate place for memfault chunk data");
+		return -ENOMEM;
+	}
+
+	data_available = memfault_packetizer_get_chunk(data_export_nfy->data, &chunk_size);
+
+	if (data_available) {
+		data_export_nfy->chunk_number = mds_instance.chunk_number;
+		data_export_nfy->rfu = 0;
+
+		/* If chunk size is smaller than maximum chunk size then shrink the buffer to actual
+		 * packet size.
+		 */
+		(void)net_buf_simple_remove_mem(&buf, (chunk_max_size - chunk_size));
+
+		err = chunk_send(conn, &buf);
+		if (err) {
+			memfault_packetizer_abort();
+			LOG_WRN("Failed to send Memfault diagnostic chunk, err %d", err);
+
+			return err;
+		}
+
+		LOG_DBG("Memfault diagnostic data chunk %d successfully sent",
+			mds_instance.chunk_number);
+		mds_instance.chunk_number = chunk_number_update(mds_instance.chunk_number);
+
+		return chunk_size;
+	}
+
+	return 0;
+}
+
+static void mds_work_handler(struct k_work *work)
+{
+	int err;
+	bool data_available;
+
+	if (!atomic_test_bit(&mds_instance.stream_state, STREAM_ENABLED)) {
+		return;
+	}
+
+	err = mds_data_send(mds_instance.conn);
+	if (err > 0) {
+		atomic_dec(&mds_instance.send_cnt);
+	}
+
+	data_available = memfault_packetizer_data_available();
+	if (data_available) {
+		if (atomic_get(&mds_instance.send_cnt)) {
+			k_work_reschedule(&mds_work, K_NO_WAIT);
```

Comment:
Why can't we call `mds_data_send()` here directly if there still is a slot in the notification pipeline instead of rescheduling? The context is the same anyway (workqueue).

Reply:
Sending data as fast as possible isn't a main case here. I would reschedule here instead of looping to allows the other items from workqueue execute without big delay in case when pipeline would have a lot of slots Also if disconnect happens we can catch it earlier and avoid losing a few notifications. What do you think?

Reply:
Yes, that makes sense, especially catching the disconnection early. Also rescheduling can help process other BLE data if application has any. So let's keep the rescheduling here.